### ISSUE TITLE
Filter out artificial elevation gain/loss

### ISSUE THREAD

**Issue by msbarry** [graphhopper/graphhopper#2028]
#1953 improves the resolution of elevation profiles so that they don't miss real elevation changes along routes, but it also makes the profile more susceptible to noise artificially inflating total elevation gain/loss metrics.

For example, going around [this turn](https://graphhopper.com/maps/?point=48.088224%2C9.108373&point=48.089317%2C9.098081&locale=en&vehicle=racingbike&weighting=fastest&elevation=true&turn_costs=false&use_miles=false&layer=TF%20Outdoors) in the Danube Valley should be flat, but any method that relies solely on DEM data will show it going up the side of the valley and back down to the river:
![image](https://user-images.githubusercontent.com/1480504/81471186-62c2c700-91bd-11ea-8963-d666b69f7e49.png).

See other examples and comparisons to other providers in comments on #1953.

**Comment by msbarry** [comment]
A few ideas off the top of my head:
- when we smooth the profile, get the min and max elevation sample used during bilinear interpolation and use that range to guess what elevation the road should be at, assuming roads are typically built to maintain a steady grade.
- during smoothing, do polyline simplification on the 2D elevation vs. distance profile of the segment, and assign elevations to points by interpolating along that (maybe using the DEM elevation range as the tolerance for each point)

**Comment by otbutz** [comment]
I think we should check whether the gradient of the path exceeds a certain threshold at any point. This should be a fairly safe criterion, otherwise the route would no longer be passable by cars.

For reference: https://en.wikipedia.org/wiki/Baldwin_Street (gradient of 35%)

**Comment by msbarry** [comment]
We'll have to take into account what kind of road as well, I know I've done at least a few hiking trails that exceed 35% for extended periods of time https://onthegomap.com/s/i76rib2u

**Comment by msbarry** [comment]
There's also probably some low hanging fruit with the existing elevation smoothing code. From my understanding it does a windowed average of all points within 150m on the segment.  This results in a couple of issues:

- points are denser around corners and each is given equal weight, so it ends up "pulling" nearby samples toward the turn (weighted average based on distance between points might help)
- if a path is going up a hill, and there is an intersection halfway up, the windowed average on the uphill side of the intersection get "pulled up" because they don't take into account samples from lower on the hill, and points on the downhill side get "pulled down" resulting in a sharp increase in a sharp increase at that intersection
- the smoothing code changes the elevation samples as it iterates through them, which leads to elevation profiles looking different based on the direction of the way

Likely, these don't impact overall elevation gain/loss much, they just result in issues with the profile shape.

**Comment by boldtrn** [comment]
One idea off the top of my head would to filter out the noise. So we should probably look at filtering out outliers.

For example for an elevation profile that looks like this:

______/\______

We should filter out that bump.

But like this:
          __         
____/      \______

We should keep it.

So for example, if there is just one point that is different than the two surrounding points, then maybe the center point is wrong.

> This results in a couple of issues

Yes, a major issue, IIRC, is that the algorithm doesn't touch the tower nodes and I think we made some compromises there in regards to speed vs. accuracy. But in general, yes the algorithm certainly has potential for improvement :).

> I think we should check whether the gradient of the path exceeds a certain threshold at any point.

I had a similar idea in the past. And I think, back then I checked for slopes > 25%, and there weren't that many, that were obviously wrong. I mean there a lot of paths up the mountain that are that steep. So that got a bit tricky to use it as rule of thumb.

> during smoothing, do polyline simplification on the 2D elevation vs. distance profile of the segment, and assign elevations to points by interpolating along that (maybe using the DEM elevation range as the tolerance for each point)

Could you elaborate on that idea? I am not sure I am understanding it.

**Comment by msbarry** [comment]
> > during smoothing, do polyline simplification on the 2D elevation vs. distance profile of the segment, and assign elevations to points by interpolating along that (maybe using the DEM elevation range as the tolerance for each point)
> 
> Could you elaborate on that idea? I am not sure I am understanding it.

After #1953 gets merged, the last step processing a way is that we do 3D polyline simplification that throws out any point that changes the 3d polyline by less than some tolerance. We are talking about setting the elevation tolerance to 3-5m, but that still retains the "exact" elevation at each point when going around turns.  During the smoothing step, we could also run 2D polyline simplification on just the elevation profile to keep the overall shape but throw out noise in the middle.

The "DEM elevation range" idea is that each point lies somewhere between 4 neighboring elevation samples from the DEM data. You know the ground rises from the min to max neighboring value, but you don't know exactly where the rise happens, so you treat the entire possible elevation range as the tolerance for that point when doing polyline simplification or another smoothing technique.  For example, if a road at 1000m goes along the edge of a cliff, every time you sample DEM data, 1 or 2 samples are in the canyon at 800-900m and the rest are at 1000m along the rim. The profile using bilinear interpolation oscillates between 900 and 1000m, but if you simplify that keeping in mind the "range" of each point, then you can throw out all of those artificial drops.

**Comment by otbutz** [comment]
I think we need to distinguish between "noise" which for me is lots of tiny changes in elevation falsifying the accumulated ascend/descend sums and real "errors" caused by outliers in the elevation profile.

**Comment by wltrr** [comment]
I also started investigating this today again as I was rather unhappy with the elevation profiles I got. I made a comparison of some data sources. The first route I considered was the street parallel to the danube above. I think it has a very similar problem to  #3031. That is, if the route travels along a steep cliff and the discretization of the dem is not parallel to the route (and cliff), a lot of noise is accumulated through the discretization:

Track on OSM Map:

<img width="1024" height="1120" alt="Image" src="https://github.com/user-attachments/assets/7ad27fd2-6f3b-444a-a12e-f853d0c5be0b" />

Track on SRTM from Kurviger (3 Arc seconds)

<img width="1024" height="1120" alt="Image" src="https://github.com/user-attachments/assets/7ca504e4-9319-4753-b0e1-5def7226bea7" />

EU-DEM from [OpenDem](https://www.opendem.info/opendemeu_background.html) (1 Arc seconds)

<img width="1024" height="1120" alt="Image" src="https://github.com/user-attachments/assets/9b903f7f-e0bf-41ba-9022-571b0623a10a" />

DE-DTM [OpenDem](https://www.opendem.info/opendtm_de.html) (1m) 

<img width="1024" height="1120" alt="Image" src="https://github.com/user-attachments/assets/a1da338b-08d4-424a-8dad-3730982584cb" />

EU-DTM from [MapsForEurope](https://www.mapsforeurope.org/datasets/euro-dem) (2 Arc second)

<img width="1024" height="1120" alt="Image" src="https://github.com/user-attachments/assets/e2d83a5c-10ab-471e-8bb1-bdb308de1ab9" />

BW-DTM [OpenGeoData-BW](https://opengeodata.lgl-bw.de/#/(sidenav:product/dgm1)) (1m resolution)

<img width="1024" height="1120" alt="Image" src="https://github.com/user-attachments/assets/9d1a11da-ddc8-453f-861d-3c0cd319a3cd" />

All images were scaled so that black is 580m and white is 750m.

Note that the last three are digital terrain models, they thus are not raw satellite signals but already try to remove buildings and canopy to directly estimate terrain height. They also seem to smooth the elevation around the street. 

Using a DTM  is even more important in the city. For example, consider this route past potsdamer platz ([graphhopper link](https://graphhopper.com/maps/point=52.510608%2C13.373501&point=52.505384%2C13.374016&profile=bike&layer=Omniscale)). Graphhopper says we go up 17m and go down 20m. This is exclusively because the srtm data does not distinguish between hills and buildings:

<img width="1024" height="1120" alt="Image" src="https://github.com/user-attachments/assets/e2731c1b-7b9a-4eaf-aed2-5101f5c05a33" />

SRTM (3 Arc seconds)

<img width="1024" height="1120" alt="Image" src="https://github.com/user-attachments/assets/1d0e5e99-11c9-498d-8890-70305b72dd92" />

EU-DTM (2 Arc Seconds)

<img width="1024" height="1120" alt="Image" src="https://github.com/user-attachments/assets/08497036-479d-4cc1-b915-c51382f15874" />

EU-DEM (1 Arc Seconds)

<img width="1024" height="1120" alt="Image" src="https://github.com/user-attachments/assets/716d59ff-cbe7-450a-bc35-dac696b29fa7" />


This time, black is 40m and white is 60m.

In my opinion EU-DEM and EU-DTM seem to be the best step forward. Their data is quite a step up from SRTM while still covering a large continuent. EU-DTM of course will be quite a lot better in cities. EU-DEM might be slightly better in the countryside.

**Comment by karussell** [comment]
> In my opinion EU-DEM and EU-DTM seem to be the best step forward.

Isn't EU-DTM including EU-DEM? From what I understand it used several data sets like EU-DEM and machine learning.
EU-DTM is under CC BY 4.0 and the different data sets seem to have also open licenses, so it seems we could use that.

btw: I didn't notice until now that [ASTER became public domain in 2019](https://www.jspacesystems.or.jp/ersdac/GDEM/E/). Important as EU-DEM is also a derived data set from ASTER...

**Comment by otbutz** [comment]
ASTER is also part of the data sources used to create NASADEM (#3147). I also think that EU-DEM is already superseded by COP-DEM and its GLO-30 model.

A digital terrain model (DTM) would be the best fit for routing purposes, as it excludes buildings and tree cover. These are still part of the "raw" digital surface models (DSM).

**Comment by wltrr** [comment]
I think resolution should be the decisive factor in the choice between datasets. 

- [One by the EEA](https://www.eea.europa.eu/en/datahub/datahubitem-view/d08852bc-7b5f-4835-a776-08362e2fbf4b?activeAccordion=#tab-metadata). It is not maintained anymore and superseded by the [Copernicus DEM / GLO-30 / GLO-90 ](https://dataspace.copernicus.eu/explore-data/data-collections/copernicus-contributing-missions/collections-description/COP-DEM). GLO-30 has 30m resolution and worldwide coverage. 
- Then there is [EuroDEM](https://www.mapsforeurope.org/datasets/euro-dem) (Not EU-DEM, my bad), which is actually a DTM. It has 2 arc seconds resolution. I believe it uses high quality official sources for buildings with which it does the post-processing.
- There is also [EU-DTM](https://portal.opentopography.org/raster?opentopoID=OTSDEM.092022.3035.1) which uses machine learning to predict terreain elevation at 30m resolution. None of the datasets used seem to contain building outlines. When looking at the Potsdamer Platz route, it also does not seem to remove the noise due to the buildings that well (see below).


**Mapsforeurope: Euro-Dem**
<img width="1008" height="1052" alt="Image" src="https://github.com/user-attachments/assets/c85d1a26-fb49-4792-9a4a-50e75c73758e" />

**OpenTopography: Euro-DTM**
<img width="1016" height="1042" alt="Image" src="https://github.com/user-attachments/assets/9a76b77c-f7bb-4abb-b702-1e88c09068be" />


**GLO-30 Donau**
<img width="1480" height="1580" alt="Image" src="https://github.com/user-attachments/assets/8fb06916-547d-415e-b5cf-bc528a5a28ec" />


**GLO-30 Potsdamer Platz**

<img width="1398" height="1464" alt="Image" src="https://github.com/user-attachments/assets/6bdadaa6-015d-498a-9644-58449fa8edbc" />

**GLO-30 Stelvio**

<img width="1658" height="2078" alt="Image" src="https://github.com/user-attachments/assets/a0c47821-2ae4-482a-a168-3e5b6a193500" />


**EuroDEM (DTM) Stelvio**

<img width="1732" height="1832" alt="Image" src="https://github.com/user-attachments/assets/1b5782da-f3af-4a8e-8bfb-9d6bf6425686" />



While the DTM is thus much better in cities, the higher resolution of GLO-30 definitely would be beneficial around very steep areas.  I did not find good sources for Nasadem or Aster, but I also do not know why I would expect them to be better than GLO-30. 

I thus suggest to first do a PR for a GLO-30 provider, for which a AWS Open Data mirror exists. Having the option to use a DTM like the one from www.mapsforeurope.org would be great. However there is no nice mirror at the moment and the file requires some post-processing (projection and tiling). One could of course also implement this provider and just explain the necessary manual work in the documentation. 

Let me know what you think about this, then I'll try to get to work soon. If you find good data sources for ASTER or NASADEM or would prefer these for some reasons please let me know too. 

(Sorry for the very long posts with screenshots again. Let me know if you think there is a better way to share these images.)

**Comment by otbutz** [comment]
> (Sorry for the very long posts with screenshots again. Let me know if you think there is a better way to share these images.)

I very much appreciate this kind of post. They replace a gut feeling with hard data.

**Comment by karussell** [comment]
GLO-30 sounds like a good replacement for cgiar. Especially since there is a mirror (and cgiar is not cleaned up regarding buildings too, if I understand it correctly).

I wonder, will we be able to replace GMTED too or should we continue using MultiSourceElevationProvider and just use GLO30 instead of CGIAR?

**update, btw:** it seems that [mapterhorn pmtiles](https://github.com/graphhopper/graphhopper/pull/3287) includes this and data with higher resolution if available. Using zoom=12 this creates a fantastic elevation profile for the provided example with a very small incline and decline:

<img width="922" height="757" alt="Image" src="https://github.com/user-attachments/assets/0263d2f2-6b4f-40db-899d-6584a1a1de66" />

(only the ledge at 48.09424,9.103269 seems to leak slightly into it)
